internal/cmd/profile/login_command.go

```diff
@@ -58,75 +60,89 @@ func loginAction(ctx *cli.Context) error {
 
 	reader := bufio.NewReader(os.Stdin)
 
-	fmt.Print("Enter Spacelift endpoint (eg. https://unicorn.app.spacelift.io/): ")
+	endpoint, err := readEndpoint(ctx, reader)
+	if err != nil {
+		return err
+	}
+	storedCredentials.Endpoint = endpoint
 
-	endpoint, err := reader.ReadString('\n')
+	credentialsType, err := getCredentialsType(ctx)
 	if err != nil {
-		return fmt.Errorf("could not read Spacelift endpoint: %w", err)
+		return err
 	}
-	endpoint = strings.TrimSpace(endpoint)
+	storedCredentials.Type = credentialsType
 
-	if endpoint == "" {
-		return errors.New("Spacelift endpoint cannot be empty")
+	switch storedCredentials.Type {
+	case session.CredentialsTypeAPIKey:
+		if err := loginUsingAPIKey(reader, &storedCredentials); err != nil {
+			return err
+		}
+	case session.CredentialsTypeGitHubToken:
+		if err := loginUsingGitHubAccessToken(&storedCredentials); err != nil {
+			return err
+		}
+	case session.CredentialsTypeAPIToken:
+		return loginUsingWebBrowser(ctx, &storedCredentials)
```

Comment:
Unless an error occurs, other methods continue to `return persistAccessCredentials(&storedCredentials)`. Why do we exit here?

Reply:
This one is a bit complicated as it opens a browser, starts a server etc. so it parses and saves the credentials in there as well. I don' think it's worth to refactor as it would make it more DRY but also more confusing.